### PR TITLE
Fixed all of logging tests

### PR DESCRIPTION
**Pull request by honza-kasik**
Ready to be merged...

### CODE REVIEW

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/ConsoleHandlerTestCase.java [hunk 1]
```diff
@@ -24,34 +14,41 @@
...
+    private static final Address C_HANDLER_TB_REMOVED_ADDRESS = LOGGING_SUBSYSTEM
+            .and("console-handler", CONSOLE_HANDLER_TB_REMOVED);
 
     @BeforeClass
-    public static void setUp() {
-        dispatcher = new Dispatcher();
-        verifier  = new ResourceVerifier(dispatcher);
+    public static void setUp() throws CommandFailedException, InterruptedException, TimeoutException, IOException {
+        operations.add(C_HANDLER_ADDRESS);
+        administration.reloadIfRequired();
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/ConsoleHandlerTestCase.java]: Are consecutive reloads necessary or will be one reload at the end sufficient?

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/ConsoleHandlerTestCase.java [hunk 2]
```diff
@@ -24,34 +14,41 @@
...
+        operations.add(C_HANDLER_TB_REMOVED_ADDRESS);
+        administration.reloadIfRequired();
     }
 
     @AfterClass
-    public static void tearDown() {
-        dispatcher.close();
+    public static void tearDown() throws InterruptedException, TimeoutException, IOException, OperationException {
+        operations.removeIfExists(C_HANDLER_ADDRESS);
+        administration.reloadIfRequired();
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/ConsoleHandlerTestCase.java]: Are consecutive reloads necessary or will be one reload at the end sufficient?

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/FileHandlerTestCase.java [hunk 3]
```diff
@@ -30,25 +29,28 @@
...
+    private static final Address FILE_HANDLER_TB_REMOVED_ADDRESS = LOGGING_SUBSYSTEM
+            .and("file-handler", FILE_HANDLER_TO_BE_REMOVED);
 
     @BeforeClass
-    public static void setUp() {
-        dispatcher = new Dispatcher();
-        verifier  = new ResourceVerifier(dispatcher);
+    public static void setUp() throws IOException, TimeoutException, InterruptedException {
+        createFileHandler(FILE_HANDLER_ADDRESS, "tests_hal1.log");
+        administration.reloadIfRequired();
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/FileHandlerTestCase.java]: Are consecutive reloads necessary or will be one reload at the end sufficient?

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/FileHandlerTestCase.java [hunk 4]
```diff
@@ -30,25 +29,28 @@
...
+        createFileHandler(FILE_HANDLER_TB_REMOVED_ADDRESS, "tests_hal2.log");
+        administration.reloadIfRequired();
     }
 
     @AfterClass
-    public static void tearDown() {
-        dispatcher.close();
+    public static void tearDown() throws IOException, OperationException, TimeoutException, InterruptedException {
+        operations.removeIfExists(FILE_HANDLER_ADDRESS);
+        administration.reloadIfRequired();
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/FileHandlerTestCase.java]: Are consecutive reloads necessary or will be one reload at the end sufficient?

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/FormatterTestCase.java [hunk 5]
```diff
@@ -22,95 +14,85 @@
...
+    private static final Address FORMATTER_ADDRESS_TO_BE_REMOVED = LOGGING_SUBSYSTEM
+            .and("pattern-formatter", FORMATTER_TO_BE_REMOVED);
 
     @BeforeClass
-    public static void setUp() {
-        dispatcher = new Dispatcher();
-        verifier  = new ResourceVerifier(dispatcher);
+    public static void setUp() throws IOException, TimeoutException, InterruptedException {
+        operations.add(FORMATTER_ADDRESS, Values.empty().and("pattern", FORMATTER_PATTERN));
+        administration.reloadIfRequired();
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/FormatterTestCase.java]: Are consecutive reloads necessary or will be one reload at the end sufficient?

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/FormatterTestCase.java [hunk 6]
```diff
@@ -22,95 +14,85 @@
...
+        operations.add(FORMATTER_ADDRESS_TO_BE_REMOVED, Values.empty().and("pattern", FORMATTER_PATTERN));
+        administration.reloadIfRequired();
     }
 
     @AfterClass
-    public static void tearDown() {
-        dispatcher.close();
+    public static void afterClass() throws InterruptedException, TimeoutException, IOException, OperationException {
+        operations.removeIfExists(FORMATTER_ADDRESS);
+        administration.reloadIfRequired();
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/FormatterTestCase.java]: Are consecutive reloads necessary or will be one reload at the end sufficient?

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/LogCategoriesTestCase.java [hunk 7]
```diff
@@ -22,32 +15,39 @@
...
+    private static final Address LOGGER_ADDRESS = LOGGING_SUBSYSTEM.and("logger", LOGGER);
+    private static final Address LOGGER_TO_BE_REMOVED_ADDRESS = LOGGING_SUBSYSTEM.and("logger", LOGGER_TO_BE_REMOVED);
 
     @BeforeClass
-    public static void setUp() {
-        dispatcher = new Dispatcher();
-        verifier  = new ResourceVerifier(dispatcher);
+    public static void beforeClass() throws CommandFailedException, InterruptedException, TimeoutException, IOException {
+        client.apply(new AddLogCategory.Builder(LOGGER).build());
+        administration.reloadIfRequired();
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/LogCategoriesTestCase.java]: Are consecutive reloads necessary or will be one reload at the end sufficient?

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/LogCategoriesTestCase.java [hunk 8]
```diff
@@ -22,32 +15,39 @@
...
+        client.apply(new AddLogCategory.Builder(LOGGER_TO_BE_REMOVED).build());
+        administration.reloadIfRequired();
     }
 
     @AfterClass
-    public static void tearDown() {
-        dispatcher.close();
+    public static void tearDown() throws IOException, TimeoutException, InterruptedException, OperationException, CommandFailedException {
+        operations.removeIfExists(LOGGER_ADDRESS);
+        administration.reloadIfRequired();
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/LogCategoriesTestCase.java]: Are consecutive reloads necessary or will be one reload at the end sufficient?

**Comment by honza-kasik** [comment]
Removed unnecessary reloads from tests

**Comment by honza-kasik** [comment]
Ready to be merged, @zebrik Could you please review?

File: common/src/main/java/org/jboss/hal/testsuite/creaper/command/logging/AddLogCategory.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    public void apply(OfflineCommandContext ctx) throws Exception {
+        GroovyXmlTransform transform = GroovyXmlTransform.of(AddLogCategory.class)
+                .subtree("logging", Subtree.subsystem("logging"))
+
+                .parameter("category", category)
+                .parameter("handlers", handlers)
+                .parameter("filter", filter)
+                .parameter("level", level == null ? null : level.value())
+                .parameter("useParentHandler", useParentHandler == null ? null : String.valueOf(useParentHandler))
+                .parameter("replaceExisting", replaceExisting)
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/creaper/command/logging/AddLogCategory.java]: Really?

File: common/src/main/java/org/jboss/hal/testsuite/creaper/command/logging/AddLogCategory.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+    private boolean replaceExisting;
+
+    protected AddLogCategory(Builder builder) {
+        setBaseProperties(builder);
+        replaceExisting = builder.replaceExisting;
+    }
+
+    @Override
+    public void apply(OfflineCommandContext ctx) throws Exception {
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/creaper/command/logging/AddLogCategory.java]: Do we need this to be implemented at all or should we rather throw something like `UnsupportedOperationException` instead?

File: common/src/main/java/org/jboss/hal/testsuite/creaper/command/logging/Level.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    WARNING("WARNING"),
+    ERROR("ERROR"),
+    SEVERE("SEVERE"),
+    FATAL("FATAL"),
+    OFF("OFF");
+
+    private final String val;
+
+    /**
+     * @param val
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/creaper/command/logging/Level.java]: Please remove this javadoc.

**Comment by zebrik** [comment]
In general LGTM. Just some minor objections .. see comments.

**Comment by rhatlapa** [comment]
Just note, the creaper commands for logging are based on https://github.com/wildfly-extras/creaper/pull/27 and thus implementations of the logging creaper commands in this PR should be only temporary and updated to those in creaper when the PR to creaper is merged and new version of creaper is released.

**Comment by Ladicek** [comment]
Here's the code that will most likely go to Creaper: https://github.com/Ladicek/creaper/commit/4a2e3d3269 It's been changed a bit, hopefully for the better :-)

**Comment by Ladicek** [comment]
FYI, I just merged https://github.com/wildfly-extras/creaper/pull/47 with the logging commands. I think that a new Creaper release should be fairly soon (but not this week).

**Comment by honza-kasik** [comment]
IMHO it would be better to wait for a Creaper release with this pull request... I'll fix it then.

**Comment by honza-kasik** [comment]
Fixed it now, it depends on Creaper's future release 0.9.3.

**Comment by Ladicek** [comment]
I'll release Creaper tomorrow or on Monday.

**Comment by honza-kasik** [comment]
Ready to be merged, Creaper 0.9.3 is available in maven repository.

**Comment by zebrik** [comment]
LGTM. Thank you!

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/LoggingAbstractTestCase.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    protected static Operations operations;
+    protected static Administration administration;
+
+    @BeforeClass
+    public static void beforeClass_() {
+        client = ManagementClientProvider.createOnlineManagementClient();
+        operations = new Operations(client);
+        administration = new Administration(client);
+        LOGGING_SUBSYSTEM = client.options().isStandalone ? Address.subsystem("logging") :
+                Address.of("profile", client.options().defaultProfile).and("subsystem", "logging");
```
Comment (Ladicek) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/LoggingAbstractTestCase.java:46]: I wonder why what Creaper is doing isn't sufficient for you in this case. Can you explain a bit? (Not that what you're doing is wrong. Just interested if Creaper could help more -- covering the difference between standalone and domain to a reasonable extent is one of the Creaper's primary goals.)

Comment (honza-kasik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/LoggingAbstractTestCase.java:46]: Well, it turned out that it was my misunderstanding... I thought that if `Address` is processed with `Operations` the adjustment to domain/standalone is not done.
The reason why this happened is funny. We are using our own `ResourceVerifier` and it throws an exception with error message which contains address without adjustment and this confused me. The problem was obviously caused by something else, but I forgot about this in the end, sorry.
I "fixed" this in another pull request...

Comment (Ladicek) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/logging/LoggingAbstractTestCase.java:46]: Thanks for the explanation. The address translation is done on the `OnlineManagementClient` level, even for composite operations ("batch"), so it should work everywhere.

**Pull request merged**